tap_workday_raas/discover.py

```diff
@@ -8,39 +8,94 @@
+
+    if is_nullable:
+        schema['type'].append('null')
+
+    if is_list:
+        schema = {
+            'type': 'array',
+            'items': schema
+        }
+
+    return schema
+
+# TODO: Test this logic
+def parse_complex_type(complex_type_selectors, xsd_schema_et, ns):
+    complex_type_mapping = {}
+    for selector in complex_type_selectors:
+        complex_type = xsd_schema_et.find(selector, ns)
+        name = complex_type.attrib['name']
+        complex_type_mapping[name] = {'type': 'object', 'properties': {}}
+        for element in complex_type.findall('.//xsd:element', ns):
+            elem_name = element.attrib['name']
+            schema_type = _element_to_schema(element)
+            complex_type_mapping[name]['properties'][elem_name] = {
+                **schema_type
+            }
+
+    return complex_type_mapping
 
-def get_schema_for_report(report, username, password):
-    xsd_schema = download_xsd(report['report_url'], username, password)
-    xsd_schema_et = ElementTree.fromstring(xsd_schema)
+def generate_schema_for_report(xsd):
+    xsd_schema_et = ElementTree.fromstring(xsd)
+    ns = {'xsd': 'http://www.w3.org/2001/XMLSchema'}
 
     schema = {'type': 'object', 'properties': {}}
 
+    # The report structure is defined by two complexType elements
+    report_structure_elem_names = {'Report_EntryType', 'Report_DataType'}
+    all_complex_type_names = {e.attrib['name'] for e in xsd_schema_et.findall("./xsd:complexType", ns)}
+
+    # The set difference results in complexType elements that are used in Report_EntryType to define nested objects
+    complex_types = all_complex_type_names - report_structure_elem_names
+
+    # Compute JSON Schemas for other complexType elements which will become nested objects
+    complex_type_mapping = parse_complex_type(["./xsd:complexType[@name='{}']".format(i) for i in complex_types], xsd_schema_et, ns)
 
-    for base_elem in xsd_schema_et:
-        if base_elem.attrib['name'] == 'Report_EntryType':
-            for sequence in base_elem:
-                for elem in sequence:
-                    elem_type = elem.attrib['type'].split(':')[1]
-                    elem_name = elem.attrib['name']
+    # Iterate the 'element' elements nested under the sequence element of the Report definition's complexType
+    for elem in xsd_schema_et.findall("./xsd:complexType[@name='Report_EntryType']/xsd:sequence/xsd:element", ns):
+        elem_type = elem.attrib['type'].split(':')[1]
+        elem_name = elem.attrib['name']
 
-                    schema_type = _type_to_schema(elem_type)
+        # When elem's type attribute is a type defined as its own complexType - a nested object
+        if elem_type in complex_type_mapping:
 
-                    schema['properties'][elem_name] = {
-                        **schema_type
-                    }
+            # TODO Check if the element is a list and type it correctly
+            is_list = elem.attrib.get("maxOccurs") == 'unbounded'
```

Comment:
This again, this should probably be a function with a standard behavior and throw if something unexpected occurs.

Reply:
Done.